CorsixTH/Lua/entities/humanoids/staff.lua

```diff
@@ -154,37 +154,43 @@ function Staff:tick()
   self:updateSpeed()
 end
 
-function Staff:checkIfWaitedTooLong()
+function Staff:checkIfWaitedTooLongForRaise()
   if self.quitting_in then
     self.quitting_in = self.quitting_in - 1
-    if self.quitting_in < 0 then
+
+    local is_waiting_time_is_up = self.quitting_in < 0
+    if is_waiting_time_is_up then
       local rise_windows = self.world.ui:getWindows(UIStaffRise)
       local staff_rise_window = nil
+      local rise_amount = self.profile:getRiseAmount()
+      self.quitting_in = nil
+      self.hospital:removeMessage(self)
 
-      -- We go through all "requesting rise" windows open, to see if we need
-      -- to close them when the person is fired.
+      -- We go through all "requesting rise" windows open
+      -- to close one of them if open when request resolved.
       for i = 1, #rise_windows do
         if rise_windows[i].staff == self then
           staff_rise_window = rise_windows[i]
           break
         end
       end
+
       -- If the hospital policy is set to automatically grant wage increases, grant
       -- the requested raise instead of firing the staff member
       if self.hospital.policies.grant_wage_increase then
-        local amount = math.floor(math.max(self.profile.wage * 1.1, (self.profile:getFairWage(self.world) + self.profile.wage) / 2) - self.profile.wage)
-        self.quitting_in = nil
-        self.hospital:removeMessage(self)
-        self:increaseWage(amount)
-        return
-      end
-
+        if staff_rise_window then -- if rise window open
+          staff_rise_window:increaseSalary() -- close window and raise
+        else
+          self:increaseWage(rise_amount)
+        end
       -- else: The staff member is sacked
-      if staff_rise_window then
-        staff_rise_window:fireStaff()
-        return
+      else
+        if staff_rise_window then
+          staff_rise_window:fireStaff() -- close window and fire
+        else
+          self:fire()
+        end
       end
```

Comment:
Having to change behaviour based on the window being open is bothersome.

I think this requires a refactor across `UIStaffRise` and `Staff:checkIfWaitedTooLongForRaise()`. Ideally, the code here should be able to handle both scenarios completely itself, closing the window at the end if it exists.

1. Identify if a window exists and remember it
2. Setup the parts to be always cleared `self.message_callback = nil` and `self.quitting_in = nil`
3. Run the raise/fire determination and payment (the auto wage grant request)
4. Close the window if one was found `staff_rise_window:close()`

Another point to bear in mind is that `UIStaffRise` stores a raise amount, but here we're disregarding that and calculating that again. It may be beneficial instead to store the rise_amount in `Staff` instead. This would require an `afterLoad` in `UIStaffRise` though akin to
```lua
if old < xxx
  self.staff.rise_amount = self.rise_amount
end
```
Alternatively, we can just ask the queued message (but would require more code to go find it if it's not open).

Reply:
I totally agree with you that the change behaviour based on the window being open is bothersome. But it was there before me and when I tried to refactor that I come to that refactoring all this will potentially bring more bugs than fix them. And I decided just to stop on this. 

So I would like to leave refactoring out of this task by now. 
Although in general I like your thoughts.

Reply:
I can sort the refactor after this pr, but might have to tweak some parts in this patch.

Reply:
> I can sort the refactor after this pr, but might have to tweak some parts in this patch.

Sounds great.